python/lsst/ts/salkafka/component_producer_set.py

```diff
@@ -60,73 +91,184 @@ def __init__(self, args):
+            # all the others and exit. This will process any task that
+            # finished and then, proceed to cancel the remaining.
+            main_tasks = [self.wait_forever_task] + self.producer_tasks
 
-    def signal_handler(self):
-        self.wait_forever_task.cancel()
+            for completed in asyncio.as_completed(main_tasks):
+                try:
+                    await completed
+                except asyncio.CancelledError:
+                    print("Terminating process.")
+                except Exception:
+                    traceback.print_exc()
+                finally:
+                    break
+
+            # Time to quit. We have received a term signal or one of the
+            # child processes has failed. Send each remaining child process
+            # a TERM signal, then exit.
+            main_process = psutil.Process(os.getpid())
+            child_processes = main_process.children(recursive=True)
+
+            # Send SIGTERM regardless of what signal was received, because
+            # it is not safe to send SIGKILL to a process that uses DDS.
+            for process in child_processes:
+                print(f"Killing child process {process}.")
+                process.send_signal(signal.SIGTERM)
+
+            # Wait for processes to terminate, on a "best effort" basis,
+            # so ignore exceptions.
+            print("Waiting for child processes to finish.")
+            await asyncio.gather(*self.producer_tasks, return_exceptions=True)
+            print("Done")
 
     @classmethod
-    async def amain(cls):
-        """Parse command line arguments, then create and run a
-        `ComponentProducerSet`.
-        """
-        parser = cls.make_argument_parser()
-        args = parser.parse_args()
-        if args.wait_for_ack != "all":
-            args.wait_for_ack = int(args.wait_for_ack)
+    def create_producer_subprocess(
+        cls, *, kafka_config, component, topic_names, log_level, queue_len,
+    ):
+        """Create and run a producer for a subset of one SAL component's
+        topics.
 
-        producer_set = ComponentProducerSet(args)
-        await producer_set.done_task
+        Parameters
+        ----------
+        kafka_config : `KafkaConfig`
+            Kafka configuration.
+        component : `str`
+            Name of a SAL component for which to handle a subset of topics.
+        topic_names : `TopicNames`
+            Topic names.
+        log_level : `int`, optional
+            Log level, e.g. logging.INFO.
+        queue_len : `int`, optional
+            Length of the DDS read queue. Must be greater than or equal to
+            `salobj.domain.DDS_READ_QUEUE_LEN`, which is the default.
+        """
+        producer_set = cls(kafka_config=kafka_config, log_level=log_level)
+        asyncio.run(
+            producer_set.run_producer_subprocess(
+                component=component, topic_names=topic_names, queue_len=queue_len,
```

Comment:
This call is missing passing `kafka_config=kafka_config`...

I think you need to do:

```
        asyncio.run(
            producer_set.run_producer_subprocess(
                kafka_config=kafka_config,
                component=component,
                topic_names=topic_names,
                queue_len=queue_len,
            )
        )
```

Reply:
Good catch. In fact the error is that `run_producer_subprocess` should not take kakfa_config as an argument; the class already has an instance of it.